provider/resources.go

```diff
@@ -454,14 +456,94 @@ func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpCl
 //go:embed cmd/pulumi-resource-gcp/bridge-metadata.json
 var metadata []byte
 
+// A predicate function that always returns true.
+func always[T any](T) bool { return true }
+
+// fixEmptyLabels applies a state edit to fix
+// https://github.com/pulumi/pulumi-gcp/issues/2372.
+func fixEmptyLabels(_ context.Context, req shimv2.PlanStateEditRequest) (cty.Value, error) {
+	// programLabels holds the labels that we expect to be applied. It is the union of
+	// labels (from the resource) and defaultLabels (from the provider).
+	//
+	// programLabels may not be the full set of labels on the resource, since
+	// effective_labels can include labels read from the cloud provider.
+	programLabels := property.Map{}
+
+	// Apply default labels first.
+	if pConfig := resource.FromResourcePropertyValue(resource.NewProperty(req.ProviderConfig)); pConfig.IsMap() {
+		l := pConfig.AsMap()["defaultLabels"]
+		if l.IsMap() {
+			programLabels = l.AsMap()
+		}
+	}
+
+	// Apply labels next, allowing labels to override defaultLabels.
+	if inputs, ok := (resource.PropertyPath{"labels"}.Get(resource.NewProperty(req.NewInputs))); ok {
+		if labels := resource.FromResourcePropertyValue(inputs); labels.IsMap() {
+			for k, v := range labels.AsMap() {
+				programLabels[k] = v
+			}
+		}
+	}
+
+	fixMap := func(f func(map[string]cty.Value) map[string]cty.Value) func(cty.Value) cty.Value {
+		return func(v cty.Value) cty.Value {
+			if !v.Type().IsMapType() || !v.IsKnown() || v.IsNull() {
+				return v
+			}
+
+			result := f(v.AsValueMap())
+			if len(result) == 0 {
+				return cty.MapValEmpty(v.Type().ElementType())
+			}
+			return cty.MapVal(result)
+		}
+	}
+
+	// fixOutputLabels fixes falsely unknown values label values.
+	fixOutputLabels := fixMap(func(m map[string]cty.Value) map[string]cty.Value {
+		for k, v := range m {
+			if v.IsKnown() {
+				// If v is known, so no correction is needed.
+				continue
+			}
+			label, ok := programLabels[k] // labels is in the Pulumi namespace
+			if !ok || label.IsComputed() || !label.IsString() {
+				// If we didn't inherit label from the program or the
+				// label is actually unknown, then just continue.
+				continue
```

Comment:
We are skipping unmanaged labels here - why does `effectiveLabels` return the correct result for them?

Reply:
We are choosing not to correct unmanaged labels here. This isn't dropping the skipped labels.

Reply:
>We are choosing not to correct unmanaged labels here

Wouldn't this mean that an empty unmanaged label causes a diff like before?

Reply:
Can we agree on a precise definition for "unmanaged" labels and keep that discoverable somewhere? Do we mean "unmanaged by Pulumi" "unmanaged by Terraform" or something else?

Reply:
>> We are choosing not to correct unmanaged labels here
> 
> Wouldn't this mean that an empty unmanaged label causes a diff like before?

Observationally they don't. I think that's because the diff customizer doesn't introduce any changes to labels imported via `pulumi refresh`.

> Can we agree on a precise definition for "unmanaged" labels and keep that discoverable somewhere?

I'm not sure where we would keep the definition. Our working definition is "unmanaged" means not managed by IaC (either TF or Pulumi).